crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -114,30 +425,123 @@ impl<S: States> Deref for State<S> {
     }
 }
 
+#[derive(Resource, Default, Debug)]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(bevy_reflect::Reflect),
+    reflect(Resource, Default)
+)]
+struct PreviousState<S: States>(S);
+
+impl<S: States> PreviousState<S> {
+    /// Get the current state.
+    pub fn get(&self) -> &S {
+        &self.0
+    }
+}
+
+impl<S: States + PartialEq> PartialEq<S> for PreviousState<S> {
+    fn eq(&self, other: &S) -> bool {
+        self.get() == other
+    }
+}
+
+impl<S: States> Deref for PreviousState<S> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        self.get()
+    }
+}
+
 /// The next state of [`State<S>`].
 ///
 /// To queue a transition, just set the contained value to `Some(next_state)`.
 /// Note that these transitions can be overridden by other systems:
 /// only the actual value of this resource at the time of [`apply_state_transition`] matters.
-#[derive(Resource, Default, Debug)]
+#[derive(Resource, Default, Clone)]
 #[cfg_attr(
     feature = "bevy_reflect",
     derive(bevy_reflect::Reflect),
     reflect(Resource, Default)
 )]
-pub struct NextState<S: States>(pub Option<S>);
+pub enum NextState<S: States> {
+    /// Do not change the state.
+    #[default]
+    Keep,
+    /// Change the state to a specific, pre-determined value
+    Value(S),
+    /// Change the state to a value determined by the given closure
+    Setter(Arc<dyn Fn(S) -> S + Sync + Send>),
+}
+
+impl<S: States> Debug for NextState<S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Keep => write!(f, "Keep"),
+            Self::Value(arg0) => f.debug_tuple("Value").field(arg0).finish(),
+            Self::Setter(_) => write!(f, "Setter"),
+        }
+    }
+}
 
 impl<S: States> NextState<S> {
     /// Tentatively set a planned state transition to `Some(state)`.
     pub fn set(&mut self, state: S) {
-        self.0 = Some(state);
+        *self = Self::Value(state);
+    }
+    /// Tentatively set a planned state transition to `Some(state)`.
+    pub fn setter(&mut self, setter: impl Fn(S) -> S + 'static + Sync + Send) {
+        *self = Self::Setter(Arc::new(setter));
     }
 }
 
+/// If the state doesn't exist, initializes it to default runs `OnEnter`
+pub fn initialize_state_and_enter<S: States>(world: &mut World) {
```

Comment:
I don't think I like this method pairs: completely removing state resources feels both expensive and somewhat confusing. Can we just cut them from this PR?

Reply:
These are there specifically for sub-states, and the concern was that we wanted to prevent invalid states, and remove the risk of the end user making mistakes setting them up.
We could make them use a `State<Option<S>>` resource instead, but that will mean that we either need to:
- make all State's contain an option so the API for sub-states and regular states is consistent
- have a completely separate set of API's for interacting with sub-states (in terms of things like OnEnter, in_state, etc.)

Alternatively, we could force the convention of having an "inactive" state on any SubStates, add a trait to codify that, and then move into and out of that inactive state.

To me, these ultimately led to the conclusion that just removing the State resource when that state shouldn't exist, and then inserting it later, is a better choice... Given these concerns, do you still feel the same way? Do you have any other suggestions on how to resolve that issue?